usr/share/rear/lib/write-protect-functions.sh

```diff
@@ -8,23 +8,72 @@ function write_protected_candidate_device() {
+    # So we only need to use the direct parent device to get all UUIDs of the whole disk
+    # because the goal is to write-protect the whole disk by using all its UUIDs
+    # cf. https://github.com/rear/rear/pull/2703#issuecomment-952888484
+    local parent_device=""
+    # Older Linux distributions do not contain lsblk (e.g. SLES10)
+    # and older lsblk versions do not support the output column PKNAME
+    # e.g. lsblk in util-linux 2.19.1 in SLES11 supports NAME and KNAME but not PKNAME
+    # cf. https://github.com/rear/rear/pull/2626#issuecomment-856700823
+    # We ignore lsblk failures and error messages and we skip empty lines in the output via 'awk NF'
+    # cf. https://unix.stackexchange.com/questions/274708/most-elegant-pipe-to-get-rid-of-empty-lines-you-can-think-of
+    # and https://stackoverflow.com/questions/23544804/how-awk-nf-filename-is-working
+    # (an empty line appears for a whole disk device e.g. /dev/sdb that has no PKNAME)
+    # and we use only the topmost reported PKNAME:
+    parent_device="$( lsblk -inpo PKNAME "$device" 2>/dev/null | awk NF | head -n1 )"
+    # parent_device is empty when lsblk does not support PKNAME.
+    # Without quoting an empty parent_device would result plain "test -b" which would (falsely) succeed:
+    test -b "$parent_device" && device="$parent_device"
 
-    if [[ " ${WRITE_PROTECTED_PARTITION_TABLE_UUIDS[*]} " == *" $partition_table_uuid "* ]]; then
-        Log "$device is designated as write-protected by partition table UUID '$partition_table_uuid'"
-        return 0
-    fi
+    local column
+    # Older lsblk versions do not support all output columns UUID PTUUID PARTUUID
+    # e.g. lsblk in util-linux 2.19.1 in SLES11 only supports UUID but neither PTUUID no PARTUUID
+    # cf. https://github.com/rear/rear/pull/2626#issuecomment-856700823
+    # When an unsupported output column is specified lsblk aborts with "unknown column" error message
+    # without output for supported output columns so we run lsblk for each output column separately
+    # and ignore lsblk failures and error messages and we skip empty lines in the output via 'awk NF'
+    # (empty lines appear when a partition does not have a filesystem UUID or for the whole device that has no PARTUUID)
+    # and we remove duplicate reported UUIDs (in particular PTUUID is reported also for each partition):
+    for column in UUID PTUUID PARTUUID ; do lsblk -ino $column "$device" 2>/dev/null ; done | awk NF | sort -u
+}
+
+function is_write_protected_by_uuid() {
+    local device="$(write_protected_candidate_device "$1")"
+    # returns 0 if one of the device's UUID is in the list of write-protected UUIDs.
 
+    local uuids uuid
+    uuids="$( write_protection_uuids "$device" )"
+    # uuids is a string of UUIDs separated by newline characters
+    if ! test "$uuids" ; then
+        LogPrintError "Cannot check write protection by UUID for $device (no UUID found)"
+        return 1
```

Comment:
isn't it safer to assume that the disk is protected (and thus return 0) if we can not determine whether it is instead of assuming that it is not protected?

Reply:
Thank you for spotting this weak point.
You are right.

Reply:
Done

Reply:
It is safer to assume that the disk is protected if we can not determine whether it is
but in practice that doesn't work sufficiently reliably because I have now
in the recovery system on a VM where sda and sdb have been RAID1 members
```
# lsblk -ipo NAME,UUID,PTUUID,PARTUUID,WWN
NAME        UUID                                 PTUUID                               PARTUUID    WWN
/dev/sda    8d05eb84-2de8-31d1-dfed-54b2ad592118                                                                           
/dev/sdb                                                                                                                   
/dev/sdc                                         e7e8e661-9040-48b5-8ca0-055f4c5d99cb                                      
```
so it can happen that a disk (sdb in this case) has none of UUID,PTUUID,PARTUUID,WWN
and with current code I get during "rear recover" (very long line wrapped here)
```
Comparing disks
Ambiguous disk layout needs manual configuration (more than one disk with same size used in '/var/lib/rear/layout/disklayout.conf')
Switching to manual disk layout configuration
Using /dev/sda (same name and same size 12884901888) for recreating /dev/sda
Cannot check write protection by ID for /dev/sdb (no ID found)
Cannot use /dev/sdb (same name and same size) for recreating /dev/sdb (/dev/sdb is write-protected)
Cannot use /dev/sda (same size) for recreating /dev/sdb (/dev/sda already exists as target in /var/lib/rear/layout/disk_mappings)
Could not automap /dev/sdb (no disk with same size 12884901888 found)
Using /dev/sdc (same name and same size 1073741824) for recreating /dev/sdc
Original disk /dev/sdb does not exist (with same size) in the target system
Cannot check write protection by ID for /dev/sdb (no ID found)
No device found where to /dev/sdb could be mapped so that it will not be recreated
Current disk mapping table (source => target):
  /dev/sda => /dev/sda
  /dev/sdc => /dev/sdc
Currently unmapped disks and dependant devices will not be recreated
(unless identical disk mapping and proceeding without manual configuration):
  /dev/sdb /dev/md127 /dev/md127p1 /dev/md127p2 /dev/mapper/cr_root
  fs:/ btrfsmountedsubvol:/ btrfsmountedsubvol:/boot/grub2/i386-pc
  btrfsmountedsubvol:/boot/grub2/x86_64-efi btrfsmountedsubvol:/opt
  btrfsmountedsubvol:/srv btrfsmountedsubvol:/home btrfsmountedsubvol:/root
  btrfsmountedsubvol:/tmp btrfsmountedsubvol:/usr/local btrfsmountedsubvol:/var
UserInput -I LAYOUT_MIGRATION_CONFIRM_MAPPINGS needed in /usr/share/rear/layout/prepare/default/300_map_disks.sh line 315
Confirm or edit the disk mapping
1) Confirm disk mapping and continue 'rear recover'
2) Confirm identical disk mapping and proceed without manual configuration
3) Edit disk mapping (/var/lib/rear/layout/disk_mappings)
4) Use Relax-and-Recover shell and return back to here
5) Abort 'rear recover'
```

By default we write protect ReaR's own disk where the recovery system is and
I think it cannot happen that this disk has none of UUID,PTUUID,PARTUUID,WWN
so it should be reasonably safe to assume a disk that has
none of UUID,PTUUID,PARTUUID,WWN is a totally empty disk that is meant
to be used to recreate the system on it so it does not need to be write-protected.

Reply:
Fixed via
https://github.com/rear/rear/commit/66cef0c8efefd2c536bfebe1d86fde1958e87b4d